### ISSUE TITLE
Filters

### ISSUE THREAD

**Issue by manuelcortez** [manuelcortez/TWBlue#102]
It would be nice to have filters in TWBlue. These filters will make buffers accept or reject everything that matches the given criteria. I tought we could implement filters by the following criteria:

* Language or list of languages.
* Word or regular expression that is present or absent in a tweet.

And we could implement them as following:

* Create a new buffer where filtered objects will be added once the given criteria will be met, it means these objects will keep being loaded in their original buffers, but will be copied to the filtered buffer.
* Filter everything in the same buffer, which means we will not create a new buffer and tweets will not be displayed if they don't match the criteria.

Thoughts?

**Comment by Mohamed00** [comment]
Perhaps if a filter contain specific words, hashtags or regular 
expressions, we could provide the user with a choice of what to 
do with the filtered item.  Either ignore it entirely, or move it 
to its own buffer.

 ----- Original Message -----

**Comment by manuelcortez** [comment]
Do you mean in filter creation right?

**Comment by Mohamed00** [comment]
Yes, during filter creation, perhaps we could ask the user what 
to do when a filtered item is encountered, and also have a 
checkbox that asks if the filtered item should be kept in the 
home buffer but not read, or moved entirely to a different 
buffer.

 ----- Original Message -----

**Comment by Mohamed00** [comment]
Are filters just saving in settings and not functional yet right now? Just asking in case I found a bug.

**Comment by manuelcortez** [comment]
What you did and what happened?

**Comment by Mohamed00** [comment]
I added a word to a filter and it didn't work. However it seems to be working with the latest commits.

**Comment by Mohamed00** [comment]
It seems that in some cases, if the [filters] section is in session.conf, it breaks client filters.

**Comment by manuelcortez** [comment]
What do you mean?
What I had a while ago was that when I added a second filter, it broke the full configuration (it duplicated entries in settings' file), but I think I will fix it soon. Are you having the same issue or are you talking about a different situation?

**Comment by Mohamed00** [comment]
### Steps to reproduce.
1. Apply a filter to a buffer, and make sure  a client is muted.
2. Remove the filter from session.conf.
### Expected.
Nothing should happen.
### Actual.
The muted client will not be ignored if it sends tweets. This only affects the buffer that had a filter applied to it.

**Comment by manuelcortez** [comment]
A few questions. Are you muting or ignoring a client? And how you did remove filters from session settings? I mean are you doing so while TWBlue is opened?

**Comment by Mohamed00** [comment]
I am ignoring a client from account settings. I've removed the filters from session.conf while TW Blue was opened and while it was closed. The results are the same. I tried erasing the cache to confirm it as well.

**Comment by manuelcortez** [comment]
@Mohamed00  can you try to repeat the process about ignoring client and stuff? it should work normally now.

**Comment by Mohamed00** [comment]
Sorry for the late reply, yes, client filters seem to be working properly now.

**Comment by Mohamed00** [comment]
Bug report. It seems that if you filter for more than one word, example: filtering for 'quote of the day:', all incoming Tweets are ignored except the Tweets that match that filter.

**Comment by manuelcortez** [comment]
Did you set "Contains" or "doesn't contain" in the first radio button?

**Comment by Mohamed00** [comment]
I set it to contains.

**Comment by manuelcortez** [comment]
@Mohamed00 Can you try again?

**Comment by manuelcortez** [comment]
I have changed the label in radio buttons for filtering by word, I hope these new labels can describe better what the option will do. Any help would be greatly appreciated :)

**Comment by Mohamed00** [comment]
It seems the filters are working properly now. I tried filtering 'Firefox 57', and all Tweets mentioning it disappeared, but nothing else did.

**Comment by manuelcortez** [comment]
I think we are almost ready with this feature for now. I just will allow everyone to set a title for every filter (we will need to identify filters with titles in the filter manager though). For now the program uses numbers (thanks time.time) for every new filter, but it looks a bit strange in the filter manager when you are viewing all your filters. Something else I will change is that in buffers, we are using just the buffer names as they are coded in the cache. I mean you will see home_timeline, mentions, direct_messages, 250942149-timeline, etc. This is a limitation for now and it will be changed when I will add a translation system for buffer names to friendly names. I think tomorrow the bot will build the new snapshot with these commits.

**Comment by Mohamed00** [comment]
Yeah, they need friendly names, and the filter manager needs the ability to edit filters. Also, minor issue, but the menu items for filters do not have any shortcuts in the graphical or invisible interfaces.

**Comment by manuelcortez** [comment]
I think for this snapshot filter manager won't be able to edit filters, though this is already planned and shouldn't be so hard to code. About shortcuts, I'll add it to the GUI and probably we will need to make some invisible shortcuts more global, so we will release some keystrokes and we can add filtering stuff there. @masonasons has suggested this in another issue.
